src/devices/BrickPi3/BrickPi3.cs

```diff
@@ -0,0 +1,948 @@
+        public string GetManufacturer()
+        {
+            string retVal = string.Empty;
+            byte[] outArray = {SPIAddress, (byte)SpiMessageType.GetManufacturer,
+                  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+            byte[] reply = SpiTransferArray(outArray);
+            if (reply[3] != 0xA5)
+            {
+                throw new IOError("No SPI response");
+            }
+            else
+            {
+                for (int ndx = 4; ndx < 24; ++ndx)
+                {
+                    if (reply[ndx] != 0)
+                    {
+                        retVal += (char)reply[ndx];
+                    }
+                    else
+                    {
+                        break;
+                    }
+                }
+            }
+            return retVal;
+        }
+
+        /// <summary>
+        /// Read the 20 charactor BrcikPi3 board name
+        /// </summary>
+        /// <returns>BrcikPi3 board name string</returns>
+        public string GetBoard()
+        {
+            string retVal = string.Empty;
+            byte[] outArray = {SPIAddress, (byte)SpiMessageType.GetName,
+                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+            byte[] reply = SpiTransferArray(outArray);
+
+            if (reply[3] == 0xA5)
+            {
+                for (int ndx = 4; ndx < 24; ++ndx)
+                {
+                    if (reply[ndx] != 0)
+                    {
+                        retVal += (char)(reply[ndx]);
+                    }
+                    else
+                    {
+                        break;
+                    }
+                }
+            }
+            else
+            {
+                throw new IOError("No SPI response");
+            }
+            return retVal;
+        }
+
+        /// <summary>
+        /// Read the hardware version
+        /// </summary>
+        /// <returns>hardware version</returns>
+        public string GetHardwareVersion()
+        {
+            string retVal = string.Empty;
+            int version = SpiRead32(SpiMessageType.GetHardwareVersion);
+            retVal = $"{version / 1000000}.{(version / 1000) % 1000}.{version % 1000}";
+            return retVal;
+        }
+
+        /// <summary>
+        /// Read the 128 - bit BrcikPi3 hardware serial number
+        /// </summary>
+        /// <returns>serial number as 32 char HEX formatted string</returns>
+        public string GetId()
+        {
+            string retVal = string.Empty;
+            byte[] outArray = {SPIAddress, (byte)SpiMessageType.GetId,
+                     0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+            byte[] reply = SpiTransferArray(outArray);
+
+            if (reply[3] == 0xA5)
+            {
+                retVal = reply[4].ToString("X2") + reply[5].ToString("X2") + reply[6].ToString("X2") +
```

Comment:
this could be: `string.Join("", reply.Take(...).Select((b) => b.ToString("X2")))`